### ISSUE TITLE
Too many instances of PrismaClient

### ISSUE THREAD

**Issue by heymartinadams** [prisma/prisma-client-js#937]
## Bug description

I’m getting the classic `FATAL: remaining connection slots are reserved for non-replication superuser connections` error for my Postgres DB because there are too many `PrismaClient` instances going on.

Problem is, I’m using REST API routes (using [NextJS](https://github.com/vercel/next.js/tree/master/examples/api-routes-rest/pages)) and not [GraphQL](https://github.com/prisma/prisma/issues/2143#issuecomment-611481435), so each time there’s an incoming request to a route (and there are many), I’m instantiating a `PrismaClient`, which results in the aforementioned error.

Is there a way to avoid this?

## Environment & setup

- OS: macOS
- Database: PostgreSQL
- Node.js version: 12.6.0
- Prisma version:
```
2.11.0
```

**Comment by bbbryan14** [comment]
The solution that I've seen is create a single instance during app initialization that is used as a dependency for modules to consume. So if you're using NestJS, I would probably just create a module that's initializes prisma client, and then have that being re-used by other services.

**Comment by heymartinadams** [comment]
@bbbryan14 how, specifically, would it be done with NextJS (not NestJS) REST APIs? As far as I know, each API is called individually at various times. Even if I separate it into modules, won’t a new Prisma Client be instantiated each time an API is called?

## module
```js
// Prisma
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()

export default prisma
```

## API 1
```js
import prisma from 'lib/utils/prisma'
...
```

## API 2
```js
import prisma from 'lib/utils/prisma'
...
```

**Comment by pantharshit00** [comment]
@heymartinadams try this: https://github.com/prisma/prisma-client-js/issues/228#issuecomment-618433162

**Comment by heymartinadams** [comment]
Will give this a try, @pantharshit00!

**Comment by vladandrei0** [comment]
@heymartinadams sorry to bother u but how did u actually made it work? If I create the module in /public/lib... and import and use it inside getServerSideProps I get: 
```
error - ./node_modules/@prisma/client/runtime/index.js:26261:39
Module not found: Can't resolve 'async_hooks'
```
🙈

**Comment by Sytten** [comment]
related to https://github.com/prisma/prisma/issues/4393 maybe?

**Comment by heymartinadams** [comment]
Works like a charm, @vladandrei0.

```js
import { PrismaClient } from '@prisma/client'

// See here: https://github.com/prisma/prisma-client-js/issues/228#issuecomment-618433162
let prisma

if (process.env.ENV === 'production') {
	prisma = new PrismaClient()
}
// `stg` or `dev`
else {
	if (!global.prisma) {
		global.prisma = new PrismaClient()
	}

	prisma = global.prisma
}

export default prisma
```

**Comment by vladandrei0** [comment]
yes, all in order now. thanks. I don't even know what I did wrong the first time...

**Comment by 2color** [comment]
Hey @heymartinadams,

Just to verify that I understand the problem correctly – is this only happening when using the `next dev` development server?

**Comment by heymartinadams** [comment]
Correct, @2color. I only use `next`, not sure if the same as `next dev`. ☺️
